server/src/main/java/org/elasticsearch/action/ActionListener.java

```diff
@@ -69,6 +70,13 @@ default <T> ActionListener<T> map(CheckedFunction<T, Response, Exception> fn) {
         return new ActionListenerImplementations.MappedActionListener<>(fn, this);
     }
 
+    /**
+     * Same as {@link #map(CheckedFunction)} except that {@code fn} is expected to never throw.
+     */
+    default <T> ActionListener<T> safeMap(Function<T, Response> fn) {
+        return new ActionListenerImplementations.SafeMappedActionListener<>(fn, this);
```

Comment:
Should we fall back to `map()` if `Assertions.ENABLED == false`? Not sure if that extra try/catch is worth avoiding in prod or not.